### DISCUSSION TITLE
How to remove label star for Sequel columns that are required/NOT NULL

### DISCUSSION THREAD

**Discussion by janko** [jeremyevans/forme#105] [General]
I'm using Forme with Roda and Sequel, and Forme is automatically appending the `*` to the form label on fields whose columns have a `NOT NULL` constraint. However, I would prefer not to display it, at least not by default. Is there a way to do that? I know I can set `required: false` on the input, but so far I didn't find a way to set it as default.

**Comment by jeremyevans** [comment]
The `*` is wrapped in an `attr` tag with `title="required"`, so I'm assuming it's easy to hide in CSS.  If not, please let me know, and we can probably add support for disabling it at the plugin level.

**Reply by janko** [reply]
Thanks, the following works:

```css
abbr[title="required"] { display: none; }
```

**Comment by v-kolesnikov** [comment]
I've noticed the same thing when I styled my forms with custom CSS according to the given design.
I think it would be better not to hardcode abbr tag and allow users to enable/disable that behaviour.

**Reply by jeremyevans** [reply]
This would break backwards compatibility, so it would have to be opt-in.  However, I'm definitely open to making it opt-in.

**Comment by jeremyevans** [comment]
@janko @v-kolesnikov What do you think of this approach?  This allows turning it off easily.  I'm open to other approaches if you have ideas in this area.

```diff
diff --git a/lib/sequel/plugins/forme.rb b/lib/sequel/plugins/forme.rb
index 2f07b48..262b5ed 100644
--- a/lib/sequel/plugins/forme.rb
+++ b/lib/sequel/plugins/forme.rb
@@ -207,7 +207,7 @@ module Sequel # :nodoc:
         # is required.
         def handle_label(f)
           opts[:label] = humanize(field) unless opts.has_key?(:label)
-          opts[:label] = [opts[:label], form._tag(:abbr, {:title=>'required'}, '*')] if opts[:label] && opts[:required]
+          opts[:label] = [opts[:label], form._tag(:abbr, {:title=>'required'}, '*')] if opts[:label] && opts[:required] && obj.forme_use_required_abbr?
         end

         # Update the attributes and options for any recognized validations
@@ -492,6 +492,11 @@ module Sequel # :nodoc:
           'post'
         end

+        # Whether to set an abbr tag in labels for required inputs.
+        def forme_use_required_abbr?
+          true
+        end
+
         # Use the underscored model name as the default namespace.
         def forme_namespace
           model.send(:underscore, model.name)
diff --git a/spec/sequel_plugin_spec.rb b/spec/sequel_plugin_spec.rb
index e1aa464..24b8d82 100644
--- a/spec/sequel_plugin_spec.rb
+++ b/spec/sequel_plugin_spec.rb
@@ -133,6 +133,11 @@ describe "Forme Sequel::Model forms" do
     @b.input(:name, :required=>true, :label=>nil).must_equal '<input id="album_name" maxlength="255" name="album[name]" required="required" type="text" value="b"/>'
   end

+  it "should not add required * in label if obj.forme_use_required_abbr? is false" do
+    def @ab.forme_use_required_abbr?; false end
+    @b.input(:name, :required=>true).must_equal '<label>Name: <input id="album_name" maxlength="255" name="album[name]" required="required" type="text" value="b"/></label>'
+  end
+
   it "should include required wrapper class if required" do
     f = Forme::Form.new(@ab, :wrapper=>:li)
     f.input(:name, :required=>true).must_equal '<li class="string required"><label>Name<abbr title="required">*</abbr>: <input id="album_name" maxlength="255" name="album[name]" required="required" type="text" value="b"/></label></li>'
```

**Reply by v-kolesnikov** [reply]
Looks good for me!

**Reply by janko** [reply]
This would require defining the `#forme_use_required_abbr?` on models we want to disable this in, correct? Would it be worth adding a plugin option to set this, something like:

```rb
Sequel::Model.plugin :forme, required_abbr: true
```

I find it more natural to do this compared to reopening the `Sequel::Model` class if I want to apply it to all models.

**Reply by jeremyevans** [reply]
@janko Correct.  I didn't think disabling this would be common enough to warrant a plugin option, but I'm open to your approach.  You could have the default implementation of `forme_use_required_abbr?` look for the plugin option. Can you submit a pull request for your approach?

**Reply by jeremyevans** [reply]
I'm going to commit the patch above.  @janko I'm still open to a PR for a plugin option if you think it is important.

**Reply by jeremyevans** [reply]
@janko Just as a heads up, I plan to release 2.4.0 early next month.

**Reply by colmexdev** [reply]
I have a different situation: I'd like to change the `title` attribute (e.g. i18n or a custom title). I checked what you say and indeed `title` is hardcoded. What would be the correct way to set the custom `title` on the `<abbr>` tag?

Thanks in advance!

**Reply by jeremyevans** [reply]
There wasn't previously support for that, but I just added it in 4b012b941aa30e4ec0def04d5855c9acbc28d3f7

**Reply by colmexdev** [reply]
Just tried it out and it's working! Thank you very much @jeremyevans
